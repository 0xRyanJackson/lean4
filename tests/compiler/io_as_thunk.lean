def p (s: String) : BaseIO String := do
  match ← (IO.println s).toBaseIO with
  | .ok _ => pure s!"done with {s}"
  | .error e => pure e.toString


def main : IO PUnit := do
  let _ ← p "aaa"
  let t1 ← (p "bbb").asThunk
  let t2 ← (p "ccc").asThunk
  let _t3 ← (p "ddd").asThunk
  IO.println t2.get
  IO.println t1.get
  IO.println t2.get
