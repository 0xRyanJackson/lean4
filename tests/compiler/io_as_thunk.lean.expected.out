aaa
ccc
done with ccc
bbb
done with bbb
done with ccc
